[Bug]: [Snappi] wrong duthost may be chosen for rx/tx port for multidut snappi tests.

### Issue Description

In pfcwd_multidut_multi_node_helper.py,

```
def run_pfcwd_multi_node_test(
......
    duthost1 = snappi_extra_params.multi_dut_params.duthost1
    rx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[0]
    rx_port_id_list = [rx_port["port_id"]]
......
    detect_time_sec = get_pfcwd_detect_time(host_ans=duthost1, intf=dut_port,
                                            asic_value=rx_port['asic_value']) / 1000.0
......
```

Here the duthost1 may not be the same device used in rx_port['peer_device'], therefore will cause failure.

If there are three ports in  snappi_extra_params.multi_dut_params.multi_dut_ports:
port 1:  duthost2, Eth256
port 2:  duthost1, Eth264
port 3:  duthost2, Eth272

Calling `get_pfcwd_detect_time` may pass in the wrong port information, such as: `get_pfcwd_detect_time( duthost1, Eth256, ...)`

Issue seems to be existing in multiple files under multidut folder, not just `run_pfcwd_multi_node_test` function.



### Results you see


```
        poll_interval_sec = get_pfcwd_poll_interval(duthost1, rx_port['asic_value']) / 1000.0
>       detect_time_sec = get_pfcwd_detect_time(host_ans=duthost1, intf=dut_port,
                                                asic_value=rx_port['asic_value']) / 1000.0
E       TypeError: unsupported operand type(s) for /: 'NoneType' and 'float'

api        = <snappi_ixnetwork.snappi_api.Api object at 0x7f4a933e3190>

```

### Results you expected to see

correct dut and port should be chosen.

### Is it platform specific

generic

### Relevant log output

_No response_

### Output of `show version`

_No response_

### Attach files (if any)

_No response_

--- a/tests/common/snappi_tests/snappi_fixtures.py
+++ b/tests/common/snappi_tests/snappi_fixtures.py
@@ -630,6 +630,7 @@ def get_multidut_snappi_ports(duthosts, conn_graph_facts, fanout_graph_facts):  
                         if port["peer_port"] in asic_port_map[asic] and hostname in port['peer_device']:
                             port['asic_value'] = asic
                             port['asic_type'] = host.facts["asic_type"]
+                            port['duthost'] = host
                             ports.append(port)
         return ports
     return _get_multidut_snappi_ports

--- a/tests/snappi_tests/multidut/ecn/files/multidut_helper.py
+++ b/tests/snappi_tests/multidut/ecn/files/multidut_helper.py
@@ -51,28 +51,32 @@ def run_ecn_test(api,
     if snappi_extra_params is None:
         snappi_extra_params = SnappiTestParams()
 
-    duthost1 = snappi_extra_params.multi_dut_params.duthost1
+    # Traffic flow:
+    # tx_port (TGEN) --- ingress DUT --- egress DUT --- rx_port (TGEN)
+
     rx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[0]
-    duthost2 = snappi_extra_params.multi_dut_params.duthost2
+    egress_duthost = rx_port['duthost']
+
     tx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[1]
+    ingress_duthost = tx_port['duthost']
 
     pytest_assert(testbed_config is not None, 'Failed to get L2/3 testbed config')
 
     logger.info("Stopping PFC watchdog")
-    stop_pfcwd(duthost1, rx_port['asic_value'])
-    stop_pfcwd(duthost2, tx_port['asic_value'])
+    stop_pfcwd(egress_duthost, rx_port['asic_value'])
+    stop_pfcwd(ingress_duthost, tx_port['asic_value'])
     logger.info("Disabling packet aging if necessary")
-    disable_packet_aging(duthost1)
-    disable_packet_aging(duthost2)
+    disable_packet_aging(egress_duthost)
+    disable_packet_aging(ingress_duthost)
 
     # Configure WRED/ECN thresholds
     logger.info("Configuring WRED and ECN thresholds")
-    config_result = config_wred(host_ans=duthost1,
+    config_result = config_wred(host_ans=egress_duthost,
                                 kmin=snappi_extra_params.ecn_params["kmin"],
                                 kmax=snappi_extra_params.ecn_params["kmax"],
                                 pmax=snappi_extra_params.ecn_params["pmax"])
     pytest_assert(config_result is True, 'Failed to configure WRED/ECN at the DUT')
-    config_result = config_wred(host_ans=duthost2,
+    config_result = config_wred(host_ans=ingress_duthost,
                                 kmin=snappi_extra_params.ecn_params["kmin"],
                                 kmax=snappi_extra_params.ecn_params["kmax"],
                                 pmax=snappi_extra_params.ecn_params["pmax"])
@@ -80,20 +84,20 @@ def run_ecn_test(api,
 
     # Enable ECN marking
     logger.info("Enabling ECN markings")
-    pytest_assert(enable_ecn(host_ans=duthost1, prio=lossless_prio), 'Unable to enable ecn')
-    pytest_assert(enable_ecn(host_ans=duthost2, prio=lossless_prio), 'Unable to enable ecn')
+    pytest_assert(enable_ecn(host_ans=egress_duthost, prio=lossless_prio), 'Unable to enable ecn')
+    pytest_assert(enable_ecn(host_ans=ingress_duthost, prio=lossless_prio), 'Unable to enable ecn')
 
-    config_result = config_ingress_lossless_buffer_alpha(host_ans=duthost1,
+    config_result = config_ingress_lossless_buffer_alpha(host_ans=egress_duthost,
                                                          alpha_log2=3)
 
     pytest_assert(config_result is True, 'Failed to configure PFC threshold to 8')
-    config_result = config_ingress_lossless_buffer_alpha(host_ans=duthost2,
+    config_result = config_ingress_lossless_buffer_alpha(host_ans=ingress_duthost,
                                                          alpha_log2=3)
 
     pytest_assert(config_result is True, 'Failed to configure PFC threshold to 8')
 
     # Get the ID of the port to test
-    port_id = get_dut_port_id(dut_hostname=duthost1.hostname,
+    port_id = get_dut_port_id(dut_hostname=egress_duthost.hostname,
                               dut_port=dut_port,
                               conn_data=conn_data,
                               fanout_data=fanout_data)
@@ -166,7 +170,7 @@ def run_ecn_test(api,
                            capture_name=snappi_extra_params.packet_capture_file)
 
         logger.info("Running traffic")
-        run_traffic(duthost=duthost1,
+        run_traffic(duthost=egress_duthost,
                     api=api,
                     config=testbed_config,
                     data_flow_names=data_flow_names,

--- a/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_basic_helper.py
+++ b/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_basic_helper.py
@@ -56,35 +56,39 @@ def run_pfcwd_basic_test(api,
     if snappi_extra_params is None:
         snappi_extra_params = SnappiTestParams()
 
-    duthost1 = snappi_extra_params.multi_dut_params.duthost1
+    # Traffic flow:
+    # tx_port (TGEN) --- ingress DUT --- egress DUT --- rx_port (TGEN)
+
     rx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[0]
-    duthost2 = snappi_extra_params.multi_dut_params.duthost2
+    egress_duthost = rx_port['duthost']
+
     tx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[1]
+    ingress_duthost = tx_port["duthost"]
     pytest_assert(testbed_config is not None, 'Fail to get L2/3 testbed config')
 
-    if (duthost1.is_multi_asic):
-        enable_packet_aging(duthost1, rx_port['asic_value'])
-        enable_packet_aging(duthost2, tx_port['asic_value'])
-        start_pfcwd(duthost1, rx_port['asic_value'])
-        start_pfcwd(duthost2, tx_port['asic_value'])
+    if (egress_duthost.is_multi_asic):
+        enable_packet_aging(egress_duthost, rx_port['asic_value'])
+        enable_packet_aging(ingress_duthost, tx_port['asic_value'])
+        start_pfcwd(egress_duthost, rx_port['asic_value'])
+        start_pfcwd(ingress_duthost, tx_port['asic_value'])
     else:
-        enable_packet_aging(duthost1)
-        enable_packet_aging(duthost2)
-        start_pfcwd(duthost1)
-        start_pfcwd(duthost2)
+        enable_packet_aging(egress_duthost)
+        enable_packet_aging(ingress_duthost)
+        start_pfcwd(egress_duthost)
+        start_pfcwd(ingress_duthost)
 
     ini_stats = {}
     for prio in prio_list:
-        ini_stats.update(get_stats(duthost1, rx_port['peer_port'], prio))
+        ini_stats.update(get_stats(egress_duthost, rx_port['peer_port'], prio))
 
     # Set appropriate pfcwd loss deviation - these values are based on empirical testing
-    DEVIATION = 0.35 if duthost1.facts['asic_type'] in ["broadcom"] or \
-        duthost2.facts['asic_type'] in ["broadcom"] else 0.3
+    DEVIATION = 0.35 if egress_duthost.facts['asic_type'] in ["broadcom"] or \
+        ingress_duthost.facts['asic_type'] in ["broadcom"] else 0.3
 
-    poll_interval_sec = get_pfcwd_poll_interval(duthost1, rx_port['asic_value']) / 1000.0
-    detect_time_sec = get_pfcwd_detect_time(host_ans=duthost1, intf=dut_port,
+    poll_interval_sec = get_pfcwd_poll_interval(egress_duthost, rx_port['asic_value']) / 1000.0
+    detect_time_sec = get_pfcwd_detect_time(host_ans=egress_duthost, intf=dut_port,
                                             asic_value=rx_port['asic_value']) / 1000.0
-    restore_time_sec = get_pfcwd_restore_time(host_ans=duthost1, intf=dut_port,
+    restore_time_sec = get_pfcwd_restore_time(host_ans=egress_duthost, intf=dut_port,
                                               asic_value=rx_port['asic_value']) / 1000.0
 
     """ Warm up traffic is initially sent before any other traffic to prevent pfcwd
@@ -148,7 +152,7 @@ def run_pfcwd_basic_test(api,
 
     fin_stats = {}
     for prio in prio_list:
-        fin_stats.update(get_stats(duthost1, rx_port['peer_port'], prio))
+        fin_stats.update(get_stats(egress_duthost, rx_port['peer_port'], prio))
 
     loss_packets = 0
     for k in fin_stats.keys():

--- a/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_burst_storm_helper.py
+++ b/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_burst_storm_helper.py
@@ -51,22 +51,27 @@ def run_pfcwd_burst_storm_test(api,
     if snappi_extra_params is None:
         snappi_extra_params = SnappiTestParams()
 
-    duthost1 = snappi_extra_params.multi_dut_params.duthost1
+    # Traffic flow:
+    # tx_port (TGEN) --- ingress DUT --- egress DUT --- rx_port (TGEN)
+
     rx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[0]
     rx_port_id = rx_port["port_id"]
-    duthost2 = snappi_extra_params.multi_dut_params.duthost2
+    egress_duthost = rx_port['duthost']
+
     tx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[1]
     tx_port_id = tx_port["port_id"]
+    ingress_duthost = tx_port['duthost']
 
     pytest_assert(testbed_config is not None, 'Fail to get L2/3 testbed config')
 
-    start_pfcwd(duthost1, rx_port['asic_value'])
-    enable_packet_aging(duthost1)
-    start_pfcwd(duthost2, tx_port['asic_value'])
-    enable_packet_aging(duthost2)
-    poll_interval_sec = get_pfcwd_poll_interval(duthost1, rx_port['asic_value']) / 1000.0
-    detect_time_sec = get_pfcwd_detect_time(host_ans=duthost1, intf=dut_port, asic_value=rx_port['asic_value']) / 1000.0        # noqa: E501
-    restore_time_sec = get_pfcwd_restore_time(host_ans=duthost1, intf=dut_port, asic_value=rx_port['asic_value']) / 1000.0      # noqa: E501
+    start_pfcwd(egress_duthost, rx_port['asic_value'])
+    enable_packet_aging(egress_duthost)
+    start_pfcwd(ingress_duthost, tx_port['asic_value'])
+    enable_packet_aging(ingress_duthost)
+
+    poll_interval_sec = get_pfcwd_poll_interval(egress_duthost, rx_port['asic_value']) / 1000.0
+    detect_time_sec = get_pfcwd_detect_time(host_ans=egress_duthost, intf=rx_port['peer_port'], asic_value=rx_port['asic_value']) / 1000.0        # noqa: E501
+    restore_time_sec = get_pfcwd_restore_time(host_ans=egress_duthost, intf=rx_port['peer_port'], asic_value=rx_port['asic_value']) / 1000.0      # noqa: E501
     burst_cycle_sec = poll_interval_sec + detect_time_sec + restore_time_sec + 0.1
     data_flow_dur_sec = ceil(burst_cycle_sec * BURST_EVENTS)
     pause_flow_dur_sec = poll_interval_sec * 0.5
@@ -108,7 +113,7 @@ def run_pfcwd_burst_storm_test(api,
     __verify_results(rows=flow_stats,
                      data_flow_prefix=DATA_FLOW_PREFIX,
                      pause_flow_prefix=PAUSE_FLOW_PREFIX,
-                     duthosts=[duthost1, duthost2])
+                     duthosts=[egress_duthost, ingress_duthost])
 
 
 def __gen_traffic(testbed_config,

--- a/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_multi_node_helper.py
+++ b/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_multi_node_helper.py
@@ -67,25 +67,37 @@ def run_pfcwd_multi_node_test(api,
     if snappi_extra_params is None:
         snappi_extra_params = SnappiTestParams()
 
-    duthost1 = snappi_extra_params.multi_dut_params.duthost1
+    # Traffic flow:
+    # tx_port (TGEN) --- ingress DUT --- egress DUT --- rx_port (TGEN)
+
+    # initialize the (duthost, port) set.
+    # The final list will have all the asics which needs to be configured for PFC
+    pfcwd_to_be_configured = set()
+
     rx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[0]
     rx_port_id_list = [rx_port["port_id"]]
-    duthost2 = snappi_extra_params.multi_dut_params.duthost2
+    egress_duthost = rx_port['duthost']
+    # Add the port to the set of ports to be configured for PFC
+    pfcwd_to_be_configured.add((egress_duthost, rx_port['asic_value']))
+
     tx_port = [snappi_extra_params.multi_dut_params.multi_dut_ports[1],
                snappi_extra_params.multi_dut_params.multi_dut_ports[2]]
     tx_port_id_list = [tx_port[0]["port_id"], tx_port[1]["port_id"]]
+    # add ingress DUT into the set
+    pfcwd_to_be_configured.add((tx_port[0]['duthost'], tx_port[0]['asic_value']))
+    pfcwd_to_be_configured.add((tx_port[1]['duthost'], tx_port[1]['asic_value']))
 
     pytest_assert(testbed_config is not None, 'Fail to get L2/3 testbed config')
     num_ports = len(port_config_list)
     pytest_require(num_ports >= 3, "This test requires at least 3 ports")
 
-    start_pfcwd(duthost1, rx_port['asic_value'])
-    enable_packet_aging(duthost1)
-    start_pfcwd(duthost2, tx_port[0]['asic_value'])
-    enable_packet_aging(duthost2)
+    # Enable PFC watchdog on the rx side and tx side of the DUT without duplication.
+    for duthost, asic in pfcwd_to_be_configured:
+        start_pfcwd(duthost, asic)
+        enable_packet_aging(duthost)
 
-    poll_interval_sec = get_pfcwd_poll_interval(duthost1, rx_port['asic_value']) / 1000.0
-    detect_time_sec = get_pfcwd_detect_time(host_ans=duthost1, intf=dut_port,
+    poll_interval_sec = get_pfcwd_poll_interval(egress_duthost, rx_port['asic_value']) / 1000.0
+    detect_time_sec = get_pfcwd_detect_time(host_ans=egress_duthost, intf=rx_port['peer_port'],
                                             asic_value=rx_port['asic_value']) / 1000.0
 
     if trigger_pfcwd:

--- a/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_runtime_traffic_helper.py
+++ b/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_runtime_traffic_helper.py
@@ -48,16 +48,20 @@ def run_pfcwd_runtime_traffic_test(api,
     if snappi_extra_params is None:
         snappi_extra_params = SnappiTestParams()
 
-    duthost1 = snappi_extra_params.duthost1
-    rx_port = snappi_extra_params.rx_port
-    duthost2 = snappi_extra_params.duthost2
-    tx_port = snappi_extra_params.tx_port
+    # Traffic flow:
+    # tx_port (TGEN) --- ingress DUT --- egress DUT --- rx_port (TGEN)
+
+    rx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[0]
+    egress_duthost = rx_port['duthost']
     rx_port_id = snappi_extra_params.rx_port_id
+
+    tx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[1]
+    ingress_duthost = tx_port['duthost']
     tx_port_id = snappi_extra_params.tx_port_id
 
     pytest_assert(testbed_config is not None, 'Fail to get L2/3 testbed config')
-    stop_pfcwd(duthost1, rx_port['asic_value'])
-    stop_pfcwd(duthost2, tx_port['asic_value'])
+    stop_pfcwd(egress_duthost, rx_port['asic_value'])
+    stop_pfcwd(ingress_duthost, tx_port['asic_value'])
 
     __gen_traffic(testbed_config=testbed_config,
                   port_config_list=port_config_list,
@@ -75,7 +79,7 @@ def run_pfcwd_runtime_traffic_test(api,
 
     flow_stats = __run_traffic(api=api,
                                config=testbed_config,
-                               duthost=duthost1,
+                               duthost=egress_duthost,
                                port=rx_port,
                                all_flow_names=all_flow_names,
                                pfcwd_start_delay_sec=PFCWD_START_DELAY_SEC,

--- a/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_runtime_traffic_helper.py
+++ b/tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_runtime_traffic_helper.py
@@ -51,11 +51,11 @@ def run_pfcwd_runtime_traffic_test(api,
     # Traffic flow:
     # tx_port (TGEN) --- ingress DUT --- egress DUT --- rx_port (TGEN)
 
-    rx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[0]
+    rx_port = snappi_extra_params.rx_port
     egress_duthost = rx_port['duthost']
     rx_port_id = snappi_extra_params.rx_port_id
 
-    tx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[1]
+    tx_port = snappi_extra_params.tx_port
     ingress_duthost = tx_port['duthost']
     tx_port_id = snappi_extra_params.tx_port_id
 

